File: Pythoncode/10.正则表达式匹配.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        i = 0
        j = 0
        if len(s) == 0 and len(p) == 0:
            return True
        if len(p) == 0 and len(s) != 0:
            return False
            
        if len(p) == 1 or p[j+1] != "*": # 第 j+1 位不是 * ，所以 p[j] 和 s[j] 一定要匹配上
            if len(s) == 0 or (p[j] != "." and p[j] != s[i]): # p[j] 和 s[j] 没匹配上
                return False
            return self.isMatch(s[i+1:], p[j+1:]) # p[j] 和 s[j] 匹配上了
        else: # 第 j+1 是 *，所以 p[j] 可以不匹配 s 里的字母，或匹配 s[i], 或 s[i], s[i+1], 或 s[i], s[i+1]，s[i+2].......
            if self.isMatch(s[i:], p[j+2:]):
                return True
            while i < len(s) and (p[j] == "." or s[i] == p[j]):
                if self.isMatch(s[i+1:], p[j+2:]):
                    return True
                i += 1
            return False

File: Cppcode/10.正则表达式匹配.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> dp;

    bool isMatch(string s, string p) {
        dp = vector<vector<int>> (s.size()+1, vector<int>(p.size()+1, 0));
        for (decltype(s.size()) i = 0; i <= s.size(); i++)
        {
            for (decltype(p.size()) j = 0; j <= p.size(); j++)
            {
                helper(s.substr(0, i), p.substr(0, j));
            }
        }
        if (dp[s.size()][p.size()] == 1) return true;
        else return false;
    }    

    void helper(string s, string p) {
        if (s.size() == 0 && p.size() == 0) dp[s.size()][p.size()] = 1;
        else if (s.size() != 0 && p.size() == 0) dp[s.size()][p.size()] = -1;
        else
        {
            if (p.back() != '*')
            {
                if (s.size() == 0 || (s.back() != p.back() && p.back() != '.')) dp[s.size()][p.size()] = -1;
                else dp[s.size()][p.size()] = dp[s.size()-1][p.size()-1];
            }
            else
            {
                if (dp[s.size()][p.size()-2] == 1) dp[s.size()][p.size()] = 1; // '*' 匹配了0个
                else if (s.size() > 0 && (p[p.size()-2] == '.' || p[p.size()-2] == s.back())) dp[s.size()][p.size()] = dp[s.size()-1][p.size()]; // '*' 匹配了至少1个
                else dp[s.size()][p.size()] = -1;
            }
        }
    }
};